---
title: " Vignette illustrating the use of MOFA on single-cell multi-omics data"
author: "Ricard Argelaguet and Britta Velten"
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc: true
vignette: >
  %\VignetteIndexEntry{Example_CLL}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

This vignette show how to MOFA can be used to disentangle the heterogeneity in single-cell DNA methylation and RNA expression (scMT) data:
![](https://github.com/bioFAM/MOFA/blob/master/images/scMT.png)
Briefly, the data set consists of 87 mouse embryonic stem cells (mESCs), comprising of 16 cells cultured in ‘2i’ media, which induces a naive pluripotency state, and 71 serum-grown cells, which commits cells to a primed pluripotency state poised for cellular differentiation (Angermueller, 2016). 


# Step 1: Load data and create MOFA object
The data is stored as a `MultiAssayExperiment` object.
Notice that there are 4 views, one with normalized RNA expression assay and 3 views with DNA methylation information. Each DNA methylation view is a different genomic context (i.e. Enhancers, Promoters and CpG Islands) and each feature is an individual CpG site.
```{r, message=FALSE}
library(MOFAtools)
library(ggplot2)
```
```{r}
data("scMT_data")
scMT_data
```
First, we create the MOFA object:
```{r}
MOFAobject <- createMOFAobject(scMT_data)
```
The function `plotTilesData` can be used to obtain an overview of the structure of the data:
```{r}
plotTilesData(MOFAobject, colors=c("#31A354","#377EB8","#377EB8","#377EB8"))
```

# Step 2: Fit the MOFA model
The next step is to train the model. Internally, this is done using Python, so make sure you have the corresponding package installed (see installation instructions in the README file).

## Define options

### Define I/O options
First, we define input and output options:
```{r}
DirOptions <- list(
  "dataDir" = tempdir(), # Directory to store the input matrices as .txt files, it can be a temporary folder
  "outFile" = "/Users/ricard/scMT_model.hdf5" # Output file of the model (use hdf5 extension)
)
```

### Define model options
Next, we define model options. The most important are:  

* **numFactors**: initial number of factors (default is 25). If you want the model to automatically learn the number of factors based on minimum variance explained, set `modelOptions$numFactors` to a large value and make sure you define the `TrainOptions$dropFactorThreshold` parameter. Alternatively, if you want to learn a fixed number of factors, set `TrainOptions$dropFactorThreshold` to zero.
* **likelihoods**: likelihood for each view. Usually we recommend gaussian for continuous data, bernoulli for binary data and poisson for count data. If not provided, the model tries to guess it from the data.
* **sparsity**: do you want to use sparsity? This makes the interpretation easier so it is recommended (Default is TRUE).

```{r}
ModelOptions <- getDefaultModelOpts(MOFAobject)
ModelOptions
```

### Define training options
Next, we define training options. The most important are:  

* **maxiter**: Maximum number of iterations. Ideally set it large enough and use the convergence criteria `TrainOptions$tolerance`.
* **learnFactors**: Maximum number of iterations, ideally set it large enough and use the convergence criteria
* **tolerance**: Convergence threshold based on change in the evidence lower bound. For an exploratory run you can use a value between 1.0 and 0.1, but for a "final" model we recommend a value of 0.01.
* **DropFactorThreshold**: Threshold of fraction of variance explained to define a factor as inactive. This is the key parameter to determine the number of factors that the model will learn. Factors explaining less than 'DropFactorThreshold' percentage of variation in all views will be removed.

```{r}
TrainOptions <- getDefaultTrainOpts()
TrainOptions$seed <- 2017
TrainOptions
```

## Prepare MOFA
`prepareMOFA` internally performs a set of sanity checks, fills the `TrainOpts` and `ModelOpts` slots of the `MOFAobject` and it also creates a set of temporary files that will be loaded by the Python framework.
```{r}
MOFAobject <- prepareMOFA(MOFAobject, 
                      DirOptions = DirOptions,
                      ModelOptions = ModelOptions,
                      TrainOptions = TrainOptions
)
```

## Run MOFA
Now we are ready to train the `MOFAobject`, which is done with the function `runMOFA`. This step can take some time (around 20 min with default parameters). For illustration we provide an existing trained `MOFAobject`
```{r, eval=FALSE}
# MOFAobject <- runMOFA(MOFAobject, DirOptions, mofaPath = "/Users/ricard/anaconda2/bin/mofa")
```

```{r}
filepath <- system.file("extdata", "scMT_model.hdf5", package = "MOFAtools")
MOFAobject <- loadModel(filepath, MOFAobject)
MOFAobject
```


# Step 3: Analyse a trained MOFA model
After training, we can explore the results from MOFA. Here we provide a semi-automated pipeline to disentangle and characterize all the identified sources of variation (the factors):
![asd](https://github.com/bioFAM/MOFA/blob/master/images/workflow.png)

<p align="center"> 
<img src="https://github.com/bioFAM/MOFA/blob/master/images/logo.png" style="width: 50%; height: 50%"/>
</p>

In a nutshell, first you need to calculate the variance explained ($R^2$) by each factor in each view. This is probably the most important result that MOFA produces, it summarises the entire heterogeneity of the dataset in a single figure: 

Next, one has to characterise the individual factors as either technical or biological (i.e phenotypic) sources of variability. It can be done via three approaches: 

* **Ordination of samples in the factor space**: this reveals clusters (or gradients), similar to what is traditionally done in Principal Component Analysis.
* **Inspection of top loadings in each factor**: the loading is a measure of feature importance, so features with high loading are the ones driving the heterogeneity captured by the factor. 
* **Feature set enrichemnt analysis**: if annotations exist (e.g. gene sets for mRNA views). 

For details, please read the Methods section of the Manuscript.

## Part 1: Disentangling the heterogeneity, calculation of variance explained by each factor in each view
This is done by `calculateVarianceExplained` and `plotVarianceExplained` function. The resulting plot allows to get an overview of which factors are active in which view(s). If a factor is active in more than one view, this means that is capturing shared signal (co-variation) between features of different views.

In this data set MOFA identified 3 Factors with a minimum variance of 2%. While Factor 1 is shared across all data modalities (7% variance explained in the RNA data and between 53% and 72% in the methylation data sets), Factors 2 and 3 are active primarily in the RNA data
```{r}

# Calculate the variance decomposition values 
r2 <- calculateVarianceExplained(MOFAobject)

# Total variance explained per view by the combination of all factors
head(r2$R2Total)

# Variance explained by each factor in each view
head(r2$R2PerFactor)

# Plot the R2 values
plotVarianceExplained(MOFAobject)
```


## Part 2: Characterisation of individual factors


### Inspection of loadings
Plotting the gene loadings for Factor 1, we can see that it is enriched for naive pluripotency marker genes such as Rex1/Zpf42, Tbx3, Fbxo15 and Essrb. Hence, based on previous studies (Mohammed et al, 2016) we can hypothetise that Factor 1 captures a transition from a naive pluripotent state to a primed pluripotent states.
```{r}
# Plot all weights and highlight specific gene markers
plotWeights(MOFAobject, view="RNA expression", factor=1, nfeatures = 0, abs = T, manual=list(c("Zfp42","Esrrb","Morc1","Fbxo15","Jam2","Klf4","Tcl1","Tbx3","Tex19.1")))

# Plot top 10 genes
plotTopWeights(MOFAobject, view="RNA expression", factor=1, nfeatures = 10)
```

A similar analysis for Factor 2 reveals that it captures a second axis of differentiation from the primed pluripotency state to a differentiated state, with highest RNA loadings for known differentiation markers such as keratins and annexins
```{r}
# Plot all weights and highlight specific gene markers
plotWeights(MOFAobject, view="RNA expression", factor=2, nfeatures = 0, manual=list(c("Krt8","Cald1","Anxa5","Tagln","Ahnak","Dsp","Anxa3","Krt19")), abs=T)

# Plot top 10 genes
plotTopWeights(MOFAobject, view="RNA expression", factor=2, nfeatures = 10)

```

Finally, instead of looking at the "abstract" weights, it is useful to observe, in the original data, the heterogeneity captured by a Factor. This can be done using the `plotDataHeatmap` function.
```{r}
# Add metadata to the plot
factor1 <- sort(getFactors(MOFAobject,"1")[,1])
order_samples <- names(factor1)
df <- data.frame(
  row.names = order_samples,
  culture = MOFAobject@InputData[,order_samples]$culture,
  factor = factor1
)

plotDataHeatmap(MOFAobject, view = "RNA expression", factor = 1, features = 20, 
                show_colnames=F, show_rownames=T, transpose = T, sortSamples = T, cluster_cols=F, annotation_col=df)
```

## Ordination of samples by factors reveals the full differentiation trajectory
Samples can be visualized along factors of interest using the `plotFactorScatter` and `plotFactorBeeswarm` functions.  

In this data set, the **combination of Factor 1 and Factor 2 captured the entire differentiation trajectory from naive pluripotent cells via primed pluripotent cells to differentiated cells**. This illustrates the importance of learning continuous latent factors rather than discrete sample assignments.
```{r}
plotFactorScatter(MOFAobject, factors=c("1","2"), color_by = "culture") +
  scale_color_manual(values=c("lightsalmon","orangered3"))
```

Finally, *Factor 3* captured the cellular detection rate, a known technical covariate that represents the number of expressed genes.
```{r}
plotFactorBeeswarm(MOFAobject, factors = 3, color_by = "cellular_detection_rate")
```


## Customized analysis
For customized exploration of weights and factors, you can directly fetch the variables from the `MOFAobject using 'get' functions: `getWeights`, `getFactors` and `getTrainData`. As an example, here we do a scatterplot between Factor 3 and the true Cellular Detection Rate values
```{r}
cdr <- colMeans(MOFAobject@TrainData$`RNA expression`>0,na.rm=T)
factor3 <- getFactors(MOFAobject, factors=3, include_intercept = F)[,1]

foo <- data.frame(factor = factor3, cdr = cdr)
ggplot(foo, aes_string(x = "factor", y = "cdr")) + 
  geom_point() + xlab("Factor 3") + ylab("Cellular Detection Rate") +
  stat_smooth(method="lm") +
  theme_bw()
```


# Further functionalities 

## Imputation of missing observations
The factors can be used to impute missing data (see Methods of the publication). This is done using the `imputeMissing` function, which stores the imputed data in the `ImputedData` slot of the `MOFAobject`. It can be accessed via the `getImputedData` function:
```{r}
MOFAobject <- imputeMissing(MOFAobject)
nonImputedMethylation <- getTrainData(MOFAobject, view="Met CpG Islands")[[1]]
imputedMethylation <- getImputedData(MOFAobject, view="Met CpG Islands")[[1]]

nonImputedMethylation[1:5,1:5]
imputedMethylation[1:5,1:5]
```

## Clustering of samples based on latent factors
Samples can be clustered according to their values on the latent factors using the `clusterSamples` function.
```{r}
# Hierarchical clustering with K=2 using Factor 1
h <- clusterSamples(MOFAobject, k=2, factors=1)

# Scatterplot colored by the predicted cluster labels and shaped by the true culture conditions
plotFactorScatter(MOFAobject, factors=1:2, color_by=h$cluster, shape_by="culture")
```


